File: akka-samples/akka-sample-persistence/src/main/scala/sample/persistence/ViewExample.scala
/**
 * Copyright (C) 2009-2013 Typesafe Inc. <http://www.typesafe.com>
 */

package sample.persistence

import akka.actor._
import akka.persistence._

object ViewExample extends App {
  class ExampleProcessor extends Processor {
    override def processorId = "processor-5"

    def receive = {
      case Persistent(payload, sequenceNr) =>
        println(s"processor received ${payload} (sequence nr = ${sequenceNr})")
    }
  }

  class ExampleView extends View {
    private var numReplicated = 0

    override def processorId = "processor-5"
    override def viewId = "view-5"

    private val destination = context.actorOf(Props[ExampleDestination])
    private val channel = context.actorOf(Channel.props("channel"))

    def receive = {
      case "snap" =>
        saveSnapshot(numReplicated)
      case SnapshotOffer(metadata, snapshot: Int) =>
        numReplicated = snapshot
        println(s"view received snapshot offer ${snapshot} (metadata = ${metadata})")
      case Persistent(payload, sequenceNr) =>
        numReplicated += 1
        println(s"view received ${payload} (sequence nr = ${sequenceNr}, num replicated = ${numReplicated})")
        channel ! Deliver(Persistent(s"replicated-${payload}"), destination.path)
    }
  }

  class ExampleDestination extends Actor {
    def receive = {
      case cp @ ConfirmablePersistent(payload, sequenceNr, _) =>
        println(s"destination received ${payload} (sequence nr = ${sequenceNr})")
        cp.confirm()
    }
  }

  val system = ActorSystem("example")

  val processor = system.actorOf(Props(classOf[ExampleProcessor]))
  val view = system.actorOf(Props(classOf[ExampleView]))

  @annotation.tailrec
  def read(line: String): Unit = line match {
    case "exit" | null =>
    case "sync" =>
      view ! Update(await = false)
      read(Console.readLine())
    case "snap" =>
      view ! "snap"
      read(Console.readLine())
    case msg =>
      processor ! Persistent(msg)
      read(Console.readLine())
  }

  read(Console.readLine())
  system.shutdown()
}

File: akka-samples/akka-sample-persistence-java/src/main/java/sample/persistence/ViewExample.java
package sample.persistence;

import java.util.concurrent.TimeUnit;
import scala.concurrent.duration.Duration;

import akka.actor.*;
import akka.persistence.*;
import akka.japi.Procedure;

public class ViewExample {
    public static class ExamplePersistentActor extends UntypedPersistentActor {
        private int count = 1;
      
        @Override
        public String persistenceId() {
            return "persistentActor-5";
        }
        
        @Override
        public void onReceiveRecover(Object message) {
          if (message instanceof String) {
            count += 1;
          } else {
            unhandled(message);
          }
        }

        @Override
        public void onReceiveCommand(Object message) {
          if (message instanceof String) {
            String s = (String) message;
            System.out.println(String.format("persistentActor received %s (nr = %d)", s, count));
            persist(s + count, new Procedure<String>() {
              public void apply(String evt) {
                count += 1;
              }
            });
          } else {
            unhandled(message);
          }
        }
    }

    public static class ExampleView extends UntypedView {

        private int numReplicated = 0;

        @Override
        public String viewId() {
            return "view-5";
        }

        @Override
        public String persistenceId() {
            return "persistentActor-5";
        }

        @Override
        public void onReceive(Object message) throws Exception {
            if (message instanceof Persistent) {
                Persistent p = (Persistent)message;
                numReplicated += 1;
                System.out.println(String.format("view received %s (num replicated = %d)", p.payload(), numReplicated));
            } else if (message instanceof SnapshotOffer) {
                SnapshotOffer so = (SnapshotOffer)message;
                numReplicated = (Integer)so.snapshot();
                System.out.println(String.format("view received snapshot offer %s (metadata = %s)", numReplicated, so.metadata()));
            } else if (message.equals("snap")) {
                saveSnapshot(numReplicated);
            }
        }
    }

    public static void main(String... args) throws Exception {
        final ActorSystem system = ActorSystem.create("example");
        final ActorRef persistentActor = system.actorOf(Props.create(ExamplePersistentActor.class));
        final ActorRef view = system.actorOf(Props.create(ExampleView.class));

        system.scheduler().schedule(Duration.Zero(), Duration.create(2, TimeUnit.SECONDS), persistentActor, "scheduled", system.dispatcher(), null);
        system.scheduler().schedule(Duration.Zero(), Duration.create(5, TimeUnit.SECONDS), view, "snap", system.dispatcher(), null);
    }
}
